allennlp_models/vision/dataset_readers/vision_reader.py

```diff
@@ -234,9 +239,37 @@ def _coordinates_cache(self) -> MutableMapping[str, Tensor]:
 
         return self._coordinates_cache_instance
 
+    @property
+    def _class_probs_cache(self) -> MutableMapping[str, Tensor]:
+        if self._class_probs_cache_instance is None:
+            if self.class_probs_cache_dir is None:
+                self._class_probs_cache_instance = {}
+            else:
+                os.makedirs(self.feature_cache_dir, exist_ok=True)  # type: ignore
+                self._class_probs_cache_instance = TensorCache(
+                    os.path.join(self.feature_cache_dir, "class_probs"),  # type: ignore
+                    read_only=not self.write_to_cache,
+                )
+
+        return self._class_probs_cache_instance
+
+    @property
+    def _class_labels_cache(self) -> MutableMapping[str, Tensor]:
+        if self._class_labels_cache_instance is None:
+            if self.class_labels_cache_dir is None:
+                self._class_labels_cache_instance = {}
+            else:
+                os.makedirs(self.feature_cache_dir, exist_ok=True)  # type: ignore
+                self._class_labels_cache_instance = TensorCache(
+                    os.path.join(self.feature_cache_dir, "class_labels"),  # type: ignore
+                    read_only=not self.write_to_cache,
+                )
+
+        return self._class_labels_cache_instance
```

Comment:
We have this code in here four times. Any chance we can simplify, maybe with the help of https://docs.python.org/3/library/functools.html#functools.cached_property?

Reply:
`cached_property` isn't available in `python3.7`, I think. I've updated the code to reduce duplicate code, however.